homeassistant/components/dlna_dms/media_source.py

```diff
@@ -0,0 +1,134 @@
+"""Implementation of DLNA DMS as a media source.
+
+URIs look like "media-source://dlna_dms/<source_id>/<media_identifier>"
+
+Media identifiers can look like:
+* `/path/to/file`: slash-separated path through the Content Directory
+* `:ObjectID`: colon followed by a server-assigned ID for an object
+* `?query`: question mark followed by a query string to search for,
+    see [DLNA ContentDirectory SearchCriteria](http://www.upnp.org/specs/av/UPnP-av-ContentDirectory-v1-Service.pdf)
+    for the syntax.
+"""
+
+from __future__ import annotations
+
+from typing import cast
+
+from homeassistant.components.media_player.const import (
+    MEDIA_CLASS_CHANNEL,
+    MEDIA_CLASS_DIRECTORY,
+    MEDIA_TYPE_CHANNEL,
+    MEDIA_TYPE_CHANNELS,
+)
+from homeassistant.components.media_player.errors import BrowseError
+from homeassistant.components.media_source.error import Unresolvable
+from homeassistant.components.media_source.models import (
+    BrowseMediaSource,
+    MediaSource,
+    MediaSourceItem,
+)
+from homeassistant.core import HomeAssistant
+
+from .const import DOMAIN, LOGGER, PATH_OBJECT_ID_FLAG, ROOT_OBJECT_ID, SOURCE_SEP
+from .dms import DidlPlayMedia, DlnaDmsData
+
+
+async def async_get_media_source(hass: HomeAssistant):
+    """Set up DLNA DMS media source."""
+    LOGGER.debug("Setting up DLNA media sources")
+    return DmsMediaSource(hass)
+
+
+class DmsMediaSource(MediaSource):
+    """Provide DLNA Media Servers as media sources."""
+
+    name = "DLNA Servers"
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize DLNA source."""
+        super().__init__(DOMAIN)
+
+        self.hass = hass
+
+    def _get_domain_data(self) -> DlnaDmsData | None:
+        """Obtain this integration's domain data if it exists."""
+        if DOMAIN in self.hass.data:
+            return cast(DlnaDmsData, self.hass.data[DOMAIN])
+        return None
+
+    async def async_resolve_media(self, item: MediaSourceItem) -> DidlPlayMedia:
+        """Resolve a media item to a playable item."""
+        dms_data = self._get_domain_data()
+        if not dms_data:
+            raise Unresolvable("dlna_dms has not been configured")
```

Comment:
This is not possible. The DLNA DMS media source is only loaded if the setup succeeded.